### PR TITLE
CSE 291 Progress Report Team 7 (Task 7: State Management)

### PR DESCRIPTION
**Pull request by yuxuan3713**
This PR includes changes made on yuxuan-dev for project progress. Base on the our workflow, this PR may seem to contain a lot of debugging commits. This PR is **not** for merge to master branch, but servers as a checkpoint.

**Recap on ourworkflow:**
- yuxuan-dev includes newest code change
- yuxuan-dev-memory and yuxuan-dev-processor are both 1 commits ahead of yuxuan-dev, and are configured for mComponent and pComponent respectively.
- After code change made to yuxuan-dev, yuxuan-dev-memory and yuxuan-dev-processor are rebased. All of three branches are pushed to GitHub and pulled to Cloudlab VM, compiled and tested.

**Progress made for sys call and p2m messaging:**

- Defined new syscall “state_dummy_get” (will be removed once APIs are functioning)
- Implemented dummy syscall number in syscall64.tbl
- Implemented dummy syscall in state.c, sending P2M msg upon syscall
- Defined P2M message opcode / msg struct / reply struct
- Implemented dummy message handler, echos input and replies
- Resolved dependency issues during implementation
- Ran dummy syscall + p2m messaging successfully
- The code structure for our project is mostly established and tested. We will fill in more APIs and functions as they are implemented and replace any dummies.

**Current dummies:**
- Dummy syscall definition
- Dummy syscall implementation, sending message to mComponent
- Dummy message handler, echos message
- Dummy test, piggy backed on general.c (piggy piggy back...)
- Tested successful syscall from pComponent, messaged mComponent, echoed message, and returned echo message to pComponent. (Details available in log files)

**Note:**
**This PR is not for merge to master branch** as it contains a lot of dummies and debugging codes. However, it reflects a working layout of our project progress and an established workflow. It will be meaningful to check for the working logic of this project. We will replace any dummies and fill in new implementations as the project goes on. We will create new "mergeable" PRs once they mature and fit in master branch.

### CODE REVIEW

File: arch/x86/entry/syscalls/syscall_64.tbl [hunk 1]
```diff
@@ -130,3 +130,6 @@
 613 	common  mq_open			sys_mq_open
 614	common  mq_receive		sys_mq_receive
 615	common  mq_close		sys_mq_close
+
+# State Management
+666 common  dummy_get       sys_dummy_get
```
Comment (yiying-zhang) [inline arch/x86/entry/syscalls/syscall_64.tbl]: user proper spacing (follow previous lines of tabs)

File: kernel/sys.c [hunk 2]
```diff
@@ -22,6 +22,9 @@
 
 #include <asm/numa.h>
 
+// State Management
+#include <processor/state.c>
```
Comment (yiying-zhang) [inline kernel/sys.c]: why are you include .c file??

File: managers/memory/core.c [hunk 3]
```diff
@@ -200,6 +200,11 @@ static void thpool_worker_handler(struct thpool_worker *worker,
 		handle_p2m_drop_page_cache(hdr, buffer);
 		break;
 
+//	State Management SYSCALL
```
Comment (yiying-zhang) [inline managers/memory/core.c]: follow the above sections format and use /* ..... */

File: managers/memory/handle_state.c [hunk 4]
```diff
@@ -0,0 +1,29 @@
+//
+// Created by Yuxuan Liu on 2/16/20.
+//
```
Comment (yiying-zhang) [inline managers/memory/handle_state.c]: We'll need to add proper header with copyright info when this is ready to be merged.

File: managers/memory/handle_state.c [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+#include <lego/printk.h>
+
+#include <memory/thread_pool.h>
+
+//TODO: handle_p2m_state_dummy_get
+struct p2m_state_reply {
+    ssize_t		retval;
+};
+
+void handle_p2m_state_dummy_get(struct p2m_state_struct *payload, struct thpool_buffer *tb)
```
Comment (yiying-zhang) [inline managers/memory/handle_state.c]: later on when you have real functions, before each function you should add a comment section like
```
/*
 * simple explanation of the function
 */
```

File: managers/memory/core.c [hunk 6]
```diff
@@ -200,6 +200,11 @@ static void thpool_worker_handler(struct thpool_worker *worker,
...
+	case P2M_STATE_DUMMY_GET:
+	    handle_p2m_state_dummy_get(payload, buffer);
```
Comment (yiying-zhang) [inline managers/memory/core.c]: I think you should also pass hdr into this function, since the real handler would need info like src_nid

**Review by yiying-zhang — CHANGES_REQUESTED** [review]
Think you'll need some changes once you start building the real handler and functionalities (which you should start doing asap to still meet you schedule).

File: kernel/sys.c (hunk 2 above)
Comment (Tian-Qiu) [reply kernel/sys.c]: We got confused about function definitions in sys.c and now it seems that both pComp and mComp compile it (functions with BUG() are not accessible to any Comps) and even though mComp doesn't call any syscall, it has to know the definition in missing_syscall.c. This dependency is correctly handled in the next pr.

File: arch/x86/entry/syscalls/syscall_64.tbl (hunk 1 above)
Comment (Tian-Qiu) [reply arch/x86/entry/syscalls/syscall_64.tbl]: Fixed.

File: managers/memory/core.c (hunk 3 above)
Comment (Tian-Qiu) [reply managers/memory/core.c]: Fixed.

File: managers/memory/handle_state.c (hunk 4 above)
Comment (Tian-Qiu) [reply managers/memory/handle_state.c]: Fixed

**Pull request closed without merging**